pkg/controllers/applicationfailover/common.go

```diff
@@ -0,0 +1,79 @@
+package applicationfailover
+
+import (
+	"sync"
+	"time"
+
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+
+	workv1alpha2 "github.com/karmada-io/karmada/pkg/apis/work/v1alpha2"
+	"github.com/karmada-io/karmada/pkg/util/fedinformer/keys"
+	"github.com/karmada-io/karmada/pkg/util/helper"
+)
+
+type workloadUnhealthyMap struct {
+	sync.RWMutex
+	// key is the resource type
+	// value is also a map. Its key is the cluster where the unhealthy workload resides.
+	// Its value is the time when the unhealthy state was first observed.
+	workloadUnhealthy map[keys.ClusterWideKey]map[string]metav1.Time
+}
+
+func newWorkloadUnhealthyMap() *workloadUnhealthyMap {
+	return &workloadUnhealthyMap{
+		workloadUnhealthy: make(map[keys.ClusterWideKey]map[string]metav1.Time),
+	}
+}
+
+func (m *workloadUnhealthyMap) delete(key keys.ClusterWideKey) {
+	m.Lock()
+	defer m.Unlock()
+	delete(m.workloadUnhealthy, key)
+}
+
+func (m *workloadUnhealthyMap) set(key keys.ClusterWideKey, unHealthyClusters map[string]metav1.Time) {
+	m.Lock()
+	defer m.Unlock()
+	m.workloadUnhealthy[key] = unHealthyClusters
+}
+
+func (m *workloadUnhealthyMap) get(key keys.ClusterWideKey) map[string]metav1.Time {
+	m.RLock()
+	defer m.RUnlock()
+	return m.workloadUnhealthy[key]
+}
+
+func detectFailure(aggregatedStatusItems []workv1alpha2.AggregatedStatusItem, gracefulEvictionTasks []workv1alpha2.GracefulEvictionTask,
+	unhealthyClusters map[string]metav1.Time, tolerationSeconds *int32) (bool, []string, map[string]struct{}) {
+	var needEvictClusters []string
+	needSecondDetection := false
```

Comment:
What's the meaning of `second delection`?

Reply:
It means when controller first finds the application on one cluster is unhealthy, it will detect the application after the tolerationSeconds is reached for the second time.